File: frontend2/src/controllers/customerController.ts
import BaseController from "./BaseController";
import Coupon from "../models/Coupon";
import Customer from "../models/Customer";
import Category from "../models/Category";
import { Console } from "console";

export default class CustomerController extends BaseController {
  constructor() {
    const baseUrl: string = "customer/";
    super(baseUrl);
  }

  async addPurchase(
    coupon: Pick<Required<Coupon>, "id">,
    customer: Pick<Required<Customer>, "id">
  ) {
    try {
      const response = await this.instance.post("addPurchase", {
        coupon: {
          id: coupon.id,
        },
        customer: {
          id: customer.id,
        },
      });
      console.log(response);
    } catch (err) {
      console.error(err);
    }
  }

  async getCustomerCoupons(
    customer: Pick<Required<Customer>, "id">
  ): Promise<Coupon[]> {
    try {
      const couponsData = (
        await this.instance.get<Coupon[]>(`getCustomerCoupons/${customer.id}`)
      ).data;
      return couponsData;
    } catch (err) {
      throw err;
    }
  }

  async getCustomerCouponsByMaxPrice(
    customer: Pick<Required<Customer>, "id">,
    maxPrice: number
  ): Promise<void> {
    try {
      const couponsData = (
        await this.instance.get<Coupon[]>(
          `getCustomerCouponsByMaxPrice/${customer.id}`,
          {
            params: {
              maxPrice,
            },
          }
        )
      ).data;
      console.log(couponsData);
    } catch (err) {
      console.error(err);
    }
  }

  async getCustomerCouponsByCategory(
    customer: Pick<Required<Customer>, "id">,
    category: Pick<Required<Category>, "id">
  ): Promise<Coupon[]> {
    try {
      const couponsData = (
        await this.instance.get<Coupon[]>(
          `getCustomerCouponsByCategory/${customer.id}`,
          {
            params: {
              categoryId: category.id,
            },
          }
        )
      ).data;
      return couponsData;
    } catch (err) {
      throw err;
    }
  }

  async getCustomerDetails(
    customer: Pick<Required<Customer>, "id">
  ): Promise<void> {
    try {
      const customerData = (
        await this.instance.get<Customer>(`getCustomerDetails/${customer.id}`)
      ).data;
      console.log(customerData);
    } catch (err) {
      console.error(err);
    }
  }
}

export function getCouponsByCustomer(customerId: number) {
  const controller = new CustomerController();
  return controller
    .getCustomerCoupons({
      id: customerId,
    })
    .then((result) => result)
    .catch((err) => console.error(err));
}

export function getCouponsByCustomerByMaxPrice(
  customerId: number,
  maxPrice: number
) {
  const controller = new CustomerController();
  return controller
    .getCustomerCouponsByMaxPrice({ id: customerId }, maxPrice)
    .then((result) => result)
    .catch((err) => console.log(err));
}

export function getCouponsByCustomerByCategory(
  customerId: number,
  categoryId: number
) {
  const controller = new CustomerController();
  return controller
    .getCustomerCouponsByCategory(
      {
        id: customerId,
      },
      {
        id: categoryId,
      }
    )
    .then((result) => result)
    .catch((err) => console.log(err));
}

File: src/main/java/com/nya/mitzi/controller/CustomerController.java
package com.nya.mitzi.controller;

import com.nya.mitzi.coupon.Coupon;
import com.nya.mitzi.coupon.CouponService;
import com.nya.mitzi.couponCustomer.CouponCustomer;
import com.nya.mitzi.customer.Customer;
import com.nya.mitzi.customer.CustomerService;
import com.nya.mitzi.exception.CouponException;
import com.nya.mitzi.exception.CustomerException;
import lombok.NoArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@NoArgsConstructor
@RequestMapping("api/customer")
@RestController
public class CustomerController {

    @Autowired
    CustomerService customerService;
    @Autowired
    CouponService couponService;

    public boolean login(String email,String password){
        return true;
    }

    @PostMapping(value="/addPurchase")
    public void purchaseCoupon(@RequestBody CouponCustomer couponCustomer){
        try {
            this.couponService.purchaseCoupon(couponCustomer.getCoupon().getId(), couponCustomer.getCustomer().getId());
        } catch (CouponException | CustomerException e) {
            throw new RuntimeException(e);
        }
    }

    @GetMapping(value="/getCustomersCoupons/{customerId}")
    public List<Coupon> getCustomerCoupons(@PathVariable int customerId) throws CustomerException {
        return this.couponService.getCouponsByCustomer(customerId);
    }

    @GetMapping(value="/getCustomersCoupons/{customerId}", params = {"categoryId"})
    public List<Coupon>getCustomerCouponsByCategory(@PathVariable int customerId,@RequestParam int categoryId) throws CustomerException {
        return this.couponService.getCouponsByCustomerAndCategory(customerId,categoryId);
    }

    @GetMapping(value="/getCustomersCoupons/{customerId}", params = {"maxPrice"})
    public List<Coupon>getCustomerCouponsByMaxPrice(@PathVariable int customerId,@RequestParam double maxPrice) throws CustomerException {
        return this.couponService.getCouponsByCustomerAndMaxPrice(customerId,maxPrice);
    }

    @GetMapping(value="/getCustomerDetails/{customerId}")
    public Customer getCustomerDetails(@PathVariable int customerId){
        try {
            return this.customerService.getCustomer(customerId);
        } catch (CustomerException e) {
            throw new RuntimeException(e);
        }
    }
}
